ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.cpp

```diff
@@ -0,0 +1,235 @@
+#include "MockVnodeAndMount.hpp"
+#include "../PrjFSKext/VnodeUtilities.hpp"
+#include "KextMockUtilities.hpp"
+#include <unordered_map>
+#include <sys/errno.h>
+
+using std::string;
+
+typedef std::unordered_map<string, VnodeWeakPointer> PathToVnodeMap;
+typedef std::unordered_map<vnode_t, VnodeWeakPointer> WeakVnodeMap;
+
+static PathToVnodeMap s_vnodesByPath;
+static WeakVnodeMap s_allVnodes;
+
+MountPointer mount::Create(const char* fileSystemTypeName, fsid_t fsid, uint64_t initialInode)
+{
+    MountPointer result(new mount{});
+    assert(strlen(fileSystemTypeName) + 1 < sizeof(result->statfs.f_fstypename));
+    result->statfs.f_fsid = fsid;
+    result->nextInode = initialInode;
+    strlcpy(result->statfs.f_fstypename, fileSystemTypeName, sizeof(result->statfs.f_fstypename));
+    return result;
+}
+
+
+vnode::vnode(const MountPointer& mount) :
+    mountPoint(mount),
+    name(nullptr),
+    inode(mount->nextInode++)
+{
+}
+
+vnode::~vnode()
+{
+    assert(this->ioCount == 0);
+}
+
+VnodePointer vnode::Create(const MountPointer& mount, const char* path, vtype vnodeType)
+{
+    VnodePointer result(new vnode(mount));
+    s_allVnodes.insert(make_pair(result.get(), VnodeWeakPointer(result)));
+    result->weakSelfPointer = result;
+    result->SetPath(path);
+    result->type = vnodeType;
+    return result;
+}
+
+VnodePointer vnode::Create(const MountPointer& mount, const char* path, vtype vnodeType, uint64_t inode)
+{
+    VnodePointer result = Create(mount, path, vnodeType);
+    result->inode = inode;
+    return result;
+}
+
+void vnode::StartRecycling()
+{
+    s_vnodesByPath.erase(this->path);
+    this->path.clear();
+    this->type = VBAD;
+    this->vid++;
+    this->isRecycling = true;
```

Comment:
With the current code, will `isRecycling` ever go back to `false`?

Reply:
Currently no, but we'll probably want that feature for other tests, e.g. for verifying that our code correctly handles the vid.